modules/leavechans.cpp

```diff
@@ -0,0 +1,270 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Automaticaly leave chans and inform users on disconnect.
+ * Author: Pierre Schweitzer <pierre@reactos.org>
+ */
+
+#include <znc/IRCNetwork.h>
+#include <znc/Chan.h>
+#include <znc/User.h>
+
+class CLeaveChan {
+public:
+    CLeaveChan() {}
+
+    CLeaveChan(const CString& sLine)
+    {
+        FromString(sLine);
+    }
+
+    CLeaveChan(const CString& sLeftChan, const CString& sNotifyChan, const CString& sNotifyMsg) :
+               m_sLeftChan(sLeftChan),
+               m_sNotifyChan(sNotifyChan),
+               m_sNotifyMsg(sNotifyMsg) {
+        m_bLeftChan = false;
+    }
+
+    virtual ~CLeaveChan() {}
+
+    const CString& GetLeftChan() const { return m_sLeftChan; }
+    const CString& GetNotifyChan() const { return m_sNotifyChan; }
+    const CString& GetNotifyMsg() const { return m_sNotifyMsg; }
+    bool HasBeenLeft() const { return m_bLeftChan; }
+    void SetChanLeft() { m_bLeftChan = true; }
+    void ResetChanLeft() { m_bLeftChan = false; }
+
+    CString ToString() const
+    {
+        return m_sLeftChan + "\t" + m_sNotifyChan + "\t" + m_sNotifyMsg;
+    }
+
+    bool FromString(const CString& sLine)
+    {
+        m_sLeftChan = sLine.Token(0, false, "\t");
+        m_sNotifyChan = sLine.Token(1, false, "\t");
+        m_sNotifyMsg = sLine.Token(2, false, "\t", true);
+        m_bLeftChan = false;
+
+        return !m_sNotifyChan.empty();
+    }
+private:
+protected:
+    bool m_bLeftChan;
+    CString m_sLeftChan;
+    CString m_sNotifyChan;
+    CString m_sNotifyMsg;
+};
+
+class CLeaveChans : public CModule {
+public:
+    MODCONSTRUCTOR(CLeaveChans) {}
+
+    virtual bool OnLoad(const CString& sArgs, CString& sMessage)
+    {
+        for (MCString::iterator it = BeginNV(); it != EndNV(); it++)
+        {
+            const CString& sLine = it->second;
+            CLeaveChan* pChan = new CLeaveChan;
+
+            if (!pChan->FromString(sLine) || FindLeftChan(pChan->GetLeftChan().AsLower()))
+            {
+                delete pChan;
+            }
+            else
+            {
+                m_msChans[pChan->GetLeftChan().AsLower()] = pChan;
+            }
+        }
+
+        return true;
+    }
+
+    virtual ~CLeaveChans()
+    {
+        for (std::map<CString, CLeaveChan*>::iterator it = m_msChans.begin(); it != m_msChans.end(); it++)
+        {
+            delete it->second;
+        }
+
+        m_msChans.clear();
+    }
+
+    virtual void OnClientDisconnect()
+    {
+        if (!m_pUser->IsUserAttached())
+        {
+            for (std::map<CString, CLeaveChan*>::const_iterator it = m_msChans.begin(); it != m_msChans.end(); it++)
```

Comment:
This is really old code, before C++11 (it land before ZNC 1.X). A revamp would be doable, for sure.